### PR TITLE
[WIP] Prototype for Collaboration

### PR DESCRIPTION
**Pull request by hkirat**

### CODE REVIEW

**Comment by hkirat** [comment]
Steps to Run
 - npm run start
 - Go to src/extensions/default/bramble/lib/RemoteCommandHandler.js 
 - Set line number 200 ([this line](https://github.com/mozilla/brackets/pull/770/files#diff-ff6f9f7c4c8b8efa6a41925f0fbe77d3R200)) to 
```js
           var delta = args[0];
```
This needs to be done to avoid the problem of same events being relayed again and again causing an infinite loop
 - Open localhost:8000/src/hostel.html
 - Click on Togetherjs button on top (only once, it takes some time to load)
 - Once you get the link to share, DONT OPEN.
 - First go to src/extensions/default/bramble/lib/RemoteCommandHandler.js line 200 and revert it to
```js
           var delta = args[0][0];
```
 - Now open the shared link.
 - Change the contents of the FIRST window. Collaboration should work.

**Comment by hkirat** [comment]
Another way of testing is not following the steps above. Simply opening, starting togetherjs and sharing.
If you try this you'll notice that right now events are being re-transmitted again and again.

Eg: User 1 adds a line, User 2 receives this change, User 2 emits this change back to user 1 and so on..
Hence as soon as you do any change, the events transmit in an infinite loop.

**Comment by humphd** [comment]
Can you rebase this so the new commits don't show up as changes?

File: src/hosted.html [hunk 1]
```diff
@@ -21,7 +21,14 @@
...
 </head>
 <body>
+<script src="https://togetherjs.com/togetherjs-min.js"></script>
+<button onclick="TogetherJS(this); return false;">Start TogetherJS</button>
     <div id="bramble" class="full"></div>
     <script src="thirdparty/require.min.js" data-main="hosted.js"></script>
-</body>
+    <script>
+        TogetherJS.hub.on('data', function (delta) {
+            console.log("bhai ka aaya data" + delta);
```
Comment (humphd) [inline src/hosted.html]: Love this, what does it mean?

File: src/extensions/default/bramble/lib/RemoteCommandHandler.js [hunk 2]
```diff
@@ -194,6 +194,30 @@ define(function (require, exports, module) {
             skipCallback = true;
             CommandManager.execute("bramble.addCodeSnippet", args[0]).always(callback);
             break;
+        case "CODEMIRROR_RELAY":
+            var editor = EditorManager.getActiveEditor();
```
Comment (humphd) [inline src/extensions/default/bramble/lib/RemoteCommandHandler.js]: We need to move all this into the editor, probably `Editor.js`, vs. doing it across the bramble `postMessage` API.

Comment (gideonthomas) [inline src/extensions/default/bramble/lib/RemoteCommandHandler.js]: +1 to this. Have a separate file in the same folder as `Editor.js` (maybe `Collaboration.js`) and put all this code in there. Then just require it in `Editor.js` and call a function to initialize it or something like that.

File: src/hosted.html (hunk 1 above)
Comment (hkirat) [inline src/hosted.html]: Yup sorry about that. Completely missed that out. 
It means "Here's your data Brother". 

Will get rid of it.

**Comment by hkirat** [comment]
Fixed the loop problem @gideonthomas. 
Working good now. The Auto inundation addon is creating a few issues, but if you turn it off it works great.

File: src/editor/Editor.js [hunk 3]
```diff
@@ -885,6 +887,18 @@ define(function (require, exports, module) {
         this._updateHiddenLines();
     };
 
+    Editor.prototype.mutualExcluse = function (f, changeList) {
```
Comment (gideonthomas) [inline src/editor/Editor.js]: what does all of this do? Please add a comment explaining.

Comment (fpbird) [inline src/editor/Editor.js]: It makes sure that if a togetherjs event has triggered a change in the codemirror instance of this particular editor, then it shouldn't transmit a new event to all the connected clients.
I saw this in y-js's collaboration code. 
But I think it should not be implemented as a prototype function to the editor, more like a global function.

Comment (hkirat) [inline src/editor/Editor.js]: It makes sure that if a togetherjs event has triggered a change in the codemirror instance of this particular editor, then it shouldn't transmit a new event to all the connected clients.
I saw this in y-js's collaboration code.
But I think it should not be implemented as a prototype function to the editor, more like a global function.

**Comment by flukeout** [comment]
Hey @hkirat - is this ready to test and experiment with? I'll spend some time today testing it out.

**Comment by gideonthomas** [comment]
@flukeout we're still working on getting the code in the right shape.

**Comment by hkirat** [comment]
Yes @flukeout as gideon mentioned the code doesn't checkout so well. 
Though it works okay. You can play around and try if you want.

File: src/editor/Editor.js [hunk 4]
```diff
@@ -77,7 +77,8 @@ define(function (require, exports, module) {
         ValidationUtils    = require("utils/ValidationUtils"),
         ViewUtils          = require("utils/ViewUtils"),
         MainViewManager    = require("view/MainViewManager"),
-        _                  = require("thirdparty/lodash");
+        _                  = require("thirdparty/lodash"),
+        brambleEvents	   = require("bramble/BrambleEvents");
```
Comment (gideonthomas) [inline src/editor/Editor.js]: In Collaboration.js:

```js
function Collaboration() {
   ...
}

Collaboration.prototype.create = function() {}

...

exports.Collaboration = Collaboration;
```

Comment (gideonthomas) [inline src/editor/Editor.js]: In Editor.js:

```js
var Collaboration = require("editor/Collaboration");

function Editor() {
    ...
    this._codemirror = ...;
    this.collaboration = new Collaboration(this._codemirror);
}
```

Comment (gideonthomas) [inline src/editor/Editor.js]: In Editor.js:
```js
function Editor(..., collaboration) {
   this._codemirror = ...;
   collaboration.setCodemirror(this._codemirror);
```

File: src/editor/Editor.js [hunk 5]
```diff
@@ -928,6 +942,7 @@ define(function (require, exports, module) {
         // whereas the "change" event should be listened to on the document. Also the
         // Editor dispatches a change event before this event is dispatched, because
         // CodeHintManager needs to hook in here when other things are already done.
+        this.mutualExcluse(brambleEvents.triggerCodeMirrorChange, changeList);
```
Comment (gideonthomas) [inline src/editor/Editor.js]:
```js
collabinstance.triggerCodemirrorchange(changeList);
```

**Comment by hkirat** [comment]
Hey @gideonthomas 
I was working on the things we discussed yesterday. requirejs doesn't seem to load togetherjs as expected.
I dont think it loads caman as well. 
If you observe [this](https://github.com/mozilla/brackets/blob/master/src/editor/Image.js#L4) during execution, the value of caman remains undefined.

**Comment by gideonthomas** [comment]
@hkirat what's the status of this?

**Comment by hkirat** [comment]
After modifying the code as we discussed, I ran into the following problems: 
1. The link as generated by togetherjs is on the path /src/index.html. This might be because we are calling Togetherjs() from inside a js file that is loaded in index.html
I tried looking into the code of togetherjs. There is a line that uses the Location variable to determine the route of the page, however modifying this to "/src/hosted.html" made togetherjs break.
2. Togetherjs doesn't function as expected. Even in one window I see a mouse pointer copying the motion of my mouse pointer.

I think loading together.min.js at the top level director (hosted.html/hosted.js) will fix this. However I think that is against the practices followed in brackets. 
Even if we do load together.min.js in hosted.html/hosted.js, the togetherjs variable will not be accessible in Collaboration.js. 
Any ideas on how we should proceed?

**Comment by gideonthomas** [comment]
1. The link that it generates for `src/index.html` is correct, it should be for that path (for now). We can change it to be configurable later. Until then, this obviously creates a problem in terms of how to access the togetherjs link from another client. you might want to edit the hosted.js file so that we allow this - i.e. accept something via the url (like the generated togetherjs link) to use instead of `src/index.html`
2. I don't understand what you're seeing. Could you elaborate some more?

**Comment by hkirat** [comment]
As per the implementation right now, on opening brackets something like this happens: 

vid

Togetherjs seems to interpret my pointer as a foreign pointer.

I see some logs along: 
screenshot from 2017-06-05 11 41 50

**Comment by hkirat** [comment]
These are the events that happen on each click, one Send and two In events
screenshot from 2017-06-05 15 26 08
It looks like for every event sent from the page, togetherjs receives the same event twice.
This is what happens on jsfiddle on each click: 
screenshot from 2017-06-05 15 29 02

File: src/editor/Collaboration.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+    Collaboration.prototype.init = function() {
+        var self = this;
+        self.changing  = false;
+        self.togetherjs.hub.on("data", function(delta) {
+            if(!self.codemirror || self.changing) {
+             return;
+            }
+            self.changing = true;
+            var cm = this.codeMirror;
```
Comment (humphd) [inline src/editor/Collaboration.js]: `this.codeMirror` should be `self.codeMirror`

Comment (hkirat) [inline src/editor/Collaboration.js]: Ah finally.
Makes sense why it wasn't working on chrome and was working on Safari.
Thanks for pointing this out.

File: src/editor/Collaboration.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            var param = delta.text.join('\n');
+            var from = cm.posFromIndex(start);
+            var to = from;
+            cm.replaceRange(param, from, to);                
+            self.changing = false;
+        });
+    };
+
+    Collaboration.prototype.setCodemirror = function(codemirror) {
+       this.codemirror = codemirror;
```
Comment (humphd) [inline src/editor/Collaboration.js]: Move this into `init` so you don't have to the check for `self.codemirror` on 19 above.

Comment (hkirat) [inline src/editor/Collaboration.js]: Had to remove this function completely for now. Apparently togetherjs handles syncing of codemirror instances. [Reference](https://github.com/mozilla/togetherjs/blob/develop/togetherjs/forms.js#L180)
So all the code in Collaboration.js became redundant.

File: src/editor/Collaboration.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
+define(function (require, exports, module) {
+    "use strict";
+
+    var togetherjs = require("togetherjs");
+
+    function Collaboration() {
+        this.togetherjs = togetherjs;
+        var self = this;
+        window.setTimeout(function() {
```
Comment (hkirat) [inline src/editor/Collaboration.js]: Added this over here to make sure that togetherjs loads after brackets has loaded completely. 
After getting loaded, togetherjs looks for all the codemirror instances to keep in sync.
Will probably move to something like
```js
window.onload
```
or
```js
document.onload
```
later.

**Comment by hkirat** [comment]
@gideonthomas As we discussed, togetherjs is *almost* similar to the one present on the website.
To include it via requirejs, I've added a 
```js
define
```
in the beginning and removed the function call in the end and returned TogetherJS from the module.
See [this](https://github.com/hkirat/brackets/blob/6eff29c9cd9b7d2b40cecb80c750f133e034a0b9/src/thirdparty/togetherjs/together.js#L6) and [this](https://github.com/hkirat/brackets/blob/6eff29c9cd9b7d2b40cecb80c750f133e034a0b9/src/thirdparty/togetherjs/together.js#L792)
Is there some other way to load Togetherjs via require using the code directly present [here](https://togetherjs.com/togetherjs-min.js)

File: src/editor/Collaboration.js (hunk 8 above)
Comment (humphd) [inline src/editor/Collaboration.js]: Using a timeout is a bad idea, let's just use existing events, or add a new one.  What are you trying to achieve exactly?

* There are a bunch of handlers for various startup events here: https://github.com/mozilla/brackets/blob/master/src/utils/AppInit.js#L130-L147
* https://github.com/mozilla/brackets/blob/master/src/extensions/default/bramble/lib/RemoteEvents.js#L237-L273 is when we are 100% loaded and running as a usable editor, and we could easily emit an internal event here

**Comment by hkirat** [comment]
Closing this as we'll be shifting from togetherjs to other libraries as discussed with gideon.

**Pull request closed without merging**